src/app/clusters/account-login-server/account-login-server.cpp

```diff
@@ -39,18 +39,51 @@
  ******************************************************************************/
 
 #include <app/Command.h>
+#include <app/common/gen/cluster-id.h>
+#include <app/common/gen/command-id.h>
 #include <app/util/af.h>
+#include <string>
 
-bool emberAfAccountLoginClusterGetSetupPINCallback(chip::app::Command * commandObj, unsigned char *)
+bool accountLoginClusterIsUserLoggedIn(std::string requestTempAccountIdentifier, std::string requestSetupPin);
+std::string accountLoginClusterGetSetupPin(std::string requestTempAccountIdentifier, chip::EndpointId endpoint);
+
+void sendResponse(chip::app::Command * command, const char * responseSetupPin)
 {
-    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
-    emberAfSendImmediateDefaultResponse(status);
+    CHIP_ERROR err                         = CHIP_NO_ERROR;
+    chip::app::CommandPathParams cmdParams = { emberAfCurrentEndpoint(), /* group id */ 0, ZCL_ACCOUNT_LOGIN_CLUSTER_ID,
+                                               ZCL_GET_SETUP_PIN_RESPONSE_COMMAND_ID,
+                                               (chip::app::CommandPathFlags::kEndpointIdValid) };
+    chip::TLV::TLVWriter * writer          = nullptr;
+    SuccessOrExit(err = command->PrepareCommand(cmdParams));
+    VerifyOrExit((writer = command->GetCommandDataElementTLVWriter()) != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    SuccessOrExit(err = writer->PutString(chip::TLV::ContextTag(0), responseSetupPin));
+    SuccessOrExit(err = command->FinishCommand());
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "Failed to encode GetSetupPIN command. Error:%s", chip::ErrorStr(err));
+    }
+}
+
+bool emberAfAccountLoginClusterGetSetupPINCallback(chip::app::Command * command, unsigned char * tempAccountIdentifier)
+{
+    std::string tempAccountIdentifierString(reinterpret_cast<char *>(tempAccountIdentifier));
```

Comment:
This is a problem.

`tempAccountIdentifier` is not a null-terminated string.  It's just whatever data was in the TLV, which does not contain a null terminator.

Unfortunately, I don't think it's a pascal string either.  Looking at the callers they are just dumping the data into the TLV (as an octet string, actually, with length provided), but the reader end just reads out the data ptr and hands it off to this method.

Which is to say that we basically don't support string args to commands for now.  https://github.com/project-chip/connectedhomeip/pull/7963 will probably fix it...

In the meantime, this is not going to work right if we ever reach this code, and possibly we should consider using an octet string for the arg for now if we want to stand this up and have it work....

Reply:
Cool. I'll watch for the PR in the future. This cluster won't be tested until TE5.